PayloadPkg/FirmwareUpdate/FirmwareUpdate.c

```diff
@@ -847,7 +847,7 @@ AuthenticateCapsule (
     }
   }
 
-  Status    = DoRsaVerify (FwImage, Header->SignatureOffset, HASH_USAGE_PUBKEY_FWU, SignatureHdr, PubKeyHdr, NULL, NULL);
+  Status    = DoRsaVerify (FwImage, Header->SignatureOffset, HASH_USAGE_PUBKEY_FWU, SignatureHdr, PubKeyHdr, PcdGet8(PcdCompSignHashAlg), NULL, NULL);
   if (EFI_ERROR (Status)) {

```

Comment:
To get hash alg, some places PcdCompSignHashAlg and some places use auth type and others. Later we need make it more clear.

Reply:
PcdCompSignHashAlg is used for component hashes and key hash store.
For RSAVerify called from Stage1B,  PubKeyHashAlg would be PcdCompSignHashAlg as configdata verification.

In ContainerFormat, we have AuthType in the header. PubKeyHashAlg would be derived from AuthType info.

Reply:
Understand current status. The CFG data should use container format already. so we could use AuthType from its header. And we would like to use container format for capsule too later.  then we could update capsule verification at that time.

Reply:
I have a issue with this 'HashAlg' parameter itself.
We should not need it.

The caller should ONLY supply : data to be verified, signature, pub key and the key_type.
Implementation should retrieve the hash and the hash_type from the trusted store (i.e. SBL hash store) and then use it.

I discussed this with Subash and it seems the root problem is our DoHashVerify API .
That API is currently being overloaded and attempting to do multiple things.
We should separate it in two modes:
DoHashVerify (Data, Length,  HashAlg, ExpectedHash) : Generic hash verification.
DoHashVerifyUsingHashStore (Data, Length, ComponentType/KeyUsage) : Use HashStore to verify the hash.

Reply:
Agree to separate 2 usage cases for DoHashVerify to make them more clear. The component hashVerify impl could get component hash internally and call generic DoHashVerify.

Reply:
Agreed, we should not pass hashtype for PubKeyhash for ones in hashstore. We should query from hashstore for algtype. We need to think about cases for pub keys which might not be in hashstore.

Also We should use separate API's for as suggested by Sachin above.